src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,357 @@
+                // 1) There is an edit following the current edit
+                // 2) The current edit and next edit involve tokens that are located right next to
+                // each other in the file.
+                // The two above criteria are also true for the similar loops in the local functions below,
+                // AddInsertionEdit and AddDeletionEdit.
+                while (i + 1 < editIndices.Length && indexToEditKinds[editIndices[i + 1]] == SemanticTokenEditKind.Update &&
+                    editIndices[i + 1] == editIndices[i] + 1)
+                {
+                    tokensToInsert.AddRange(newGroupedSemanticTokens[editIndices[i + 1]].ConvertToArray());
+                    deleteCount += 5;
+                    i++;
+                }
+
+                semanticTokensEdits.Add(
+                    GenerateEdit(start: editStartPosition, deleteCount: deleteCount, data: tokensToInsert.ToArray()));
+
+                return i;
+            }
+
+            static int AddInsertionEdit(
+                SemanticToken[] newGroupedSemanticTokens,
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndices,
+                int i,
+                SemanticToken groupedSemanticToken,
+                int editStartPosition)
+            {
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                tokensToInsert.AddRange(groupedSemanticToken.ConvertToArray());
+
+                // An insert can only be combined with other inserts that directly follow it.
+                while (i + 1 < editIndices.Length && indexToEditKinds[editIndices[i + 1]] == SemanticTokenEditKind.Insert &&
+                    editIndices[i + 1] == editIndices[i] + 1)
+                {
+                    tokensToInsert.AddRange(newGroupedSemanticTokens[editIndices[i + 1]].ConvertToArray());
+                    i++;
+                }
+
+                semanticTokensEdits.Add(
+                    GenerateEdit(start: editStartPosition, deleteCount: 0, data: tokensToInsert.ToArray()));
+                return i;
+            }
+
+            static int AddDeletionEdit(
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndices,
+                int i,
+                int editStartPosition)
+            {
+                var deleteCount = 5;
+
+                // A deletion can only be combined with other deletions that directly follow it.
+                while (i + 1 < editIndices.Length && indexToEditKinds[editIndices[i + 1]] == SemanticTokenEditKind.Delete &&
+                    editIndices[i + 1] == editIndices[i] + 1)
+                {
+                    deleteCount += 5;
+                    i++;
+                }
```

Comment:
suggestion.  this walk is repeated in all three methods.  consider a helper that just takes in the array, and the type of edit kidn to group and it walks forward and tells you how much to skip forward.  you then call that, and use the result to inform your computations (i.e. increasing 'i' or upping your deletion count).